client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java

```diff
@@ -978,21 +979,43 @@ protected final boolean isCachingEnabled() {
 
     /**
      * Gets a response from the cache.
-     * 
+     *
      * @param key The key to get the response for.
      * @return A succeeded future containing the response from the cache
      *         or a failed future if no response exists for the key
      *         or the response is expired.
      */
     protected Future<R> getResponseFromCache(final Object key) {
+        return getResponseFromCache(key, null);
+    }
+
+    /**
+     * Gets a response from the cache.
+     * <p>
+     * Sets tags on the given span according to whether there was a cache hit and according to what the response result
+     * status is (if there was a cache hit).
+     *
+     * @param key The key to get the response for.
+     * @param currentSpan The span to mark (may be {@code null}).
+     * @return A succeeded future containing the response from the cache
+     *         or a failed future if no response exists for the key
+     *         or the response is expired.
+     */
+    protected Future<R> getResponseFromCache(final Object key, final Span currentSpan) {
 
         if (responseCache == null) {
             return Future.failedFuture(new IllegalStateException("no cache configured"));
         } else {
             final R result = responseCache.get(key);
+            if (currentSpan != null) {
+                TracingHelper.TAG_CACHE_HIT.set(currentSpan, result != null);
+            }
             if (result == null) {
                 return Future.failedFuture("cache miss");
             } else {
+                if (currentSpan != null) {
+                    Tags.HTTP_STATUS.set(currentSpan, result.getStatus());
```

Comment:
IMHO this shouldn't be done here because if there is a cache miss, the client of this method needs to retrieve the result from somewhere else and then also put the status to the span. It seems natural to me that putting the status to the span is the responsibility of the client invoking this method whereas marking the span for cache hit/miss is this method's responsibility.

Reply:
> this shouldn't be done here because if there is a cache miss, the client of this method needs to retrieve the result from somewhere else and then also put the status to the span.

The "getting the result from somewhere else" usually means getting it via a `createAndSendRequest()` invocation. Inside `createAndSendRequest()`, the status tag is already set on the span.

That means the place in the code where `getResponseFromCache()` and as fall-back `createAndSendRequest()` is invoked doesn't set the status tag as of now. The status tag is set deep down where the result is set in the Future. And following that logic, I think we should set it in `getResponseFromCache()` as well.

An alternative would be setting it on the level where the span is finished, i.e in `mapResultAndFinishSpan()` for example. But then I would remove the places where the status is set in `createAndSendRequest()` (or deeper down).
Maybe this is better, since we also log an error to the span at the level of `mapResultAndFinishSpan()` - not at the place where it is set in the Future result.

Reply:
Yeah, I think that would be a more consistent approach.

Reply:
I've added a commit implementing the aforementioned alternative, setting the status tag now where also a potential error is logged to the trace span.